## 1. Import Libraries


# Employee Attrition Prediction using Machine Learning

**Dataset:** IBM HR Analytics Employee Attrition Dataset

**Algorithm:** Logistic Regression

**Objective:** Predict whether an employee is likely to leave the company.


In [1]:

import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report


## 2. Load Dataset

In [2]:

df = pd.read_csv(r"WA_Fn-UseC_-HR-Employee-Attrition.csv")
df.head()


,Age,Attrition,BusinessTravel,DailyRate,Department,DistanceFromHome,Education,EducationField,EmployeeCount,EmployeeNumber,...,RelationshipSatisfaction,StandardHours,StockOptionLevel,TotalWorkingYears,TrainingTimesLastYear,WorkLifeBalance,YearsAtCompany,YearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager
0,41,Yes,Travel_Rarely,1102,Sales,1,2,Life Sciences,1,1,...,1,80,0,8,0,1,6,4,0,5
1,49,No,Travel_Frequently,279,Research & Development,8,1,Life Sciences,1,2,...,4,80,1,10,3,3,10,7,1,7
2,37,Yes,Travel_Rarely,1373,Research & Development,2,2,Other,1,4,...,2,80,0,7,3,3,0,0,0,0
3,33,No,Travel_Frequently,1392,Research & Development,3,4,Life Sciences,1,5,...,3,80,0,8,3,3,8,7,3,0
4,27,No,Travel_Rarely,591,Research & Development,2,1,Medical,1,7,...,4,80,1,6,3,3,2,2,2,2


## 3. Explore the Dataset

In [3]:

print(df.shape)
print(df.info())
print(df.isnull().sum())
print(df["Attrition"].value_counts())


(1470, 35)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel     

## 4. Data Preprocessing

In [4]:

df = df.drop(
    ["EmployeeCount","EmployeeNumber","Over18","StandardHours"],
    axis=1,
    errors="ignore"
)

df["Attrition"] = df["Attrition"].map({"Yes":1,"No":0})

df = pd.get_dummies(df, drop_first=True)


## 5. Train-Test Split

In [5]:

X = df.drop("Attrition", axis=1)
y = df["Attrition"]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)


## 6. Feature Scaling

In [6]:

scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


## 7. Train Logistic Regression Model

In [7]:

model = LogisticRegression(max_iter=5000)
model.fit(X_train_scaled, y_train)


LogisticRegression(max_iter=5000)

## 8. Evaluate the Model

In [8]:

y_pred = model.predict(X_test_scaled)

print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nConfusion Matrix")
print(confusion_matrix(y_test, y_pred))
print("\nClassification Report")
print(classification_report(y_test, y_pred))


Accuracy: 0.8775510204081632

Confusion Matrix
[[241  14]
 [ 22  17]]

Classification Report
              precision    recall  f1-score   support

           0       0.92      0.95      0.93       255
           1       0.55      0.44      0.49        39

    accuracy                           0.88       294
   macro avg       0.73      0.69      0.71       294
weighted avg       0.87      0.88      0.87       294



## 9. Sample Prediction

In [9]:

sample = X_test_scaled[0].reshape(1, -1)
prediction = model.predict(sample)

if prediction[0] == 1:
    print("Employee is likely to leave.")
else:
    print("Employee is likely to stay.")


Employee is likely to stay.



## 10. Conclusion

- Logistic Regression was used to predict employee attrition.
- Dataset was cleaned and preprocessed before training.
- The trained model achieved approximately **87.76% accuracy**.
- This model can assist HR departments in identifying employees who may be at risk of leaving.
